[profiler] log process group config information in distributedInfo field

shengbao-zheng opened the pull request:
Summary: Process Group config is essential to analyze collective pattern. We have added this in Execution Trace. Now expose this information in Kineto as well

Differential Revision: D53557965




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

wconstab reviewed (commented):
this looks fine, though i wonder if we have any hesitation about adding the new api without more consideration of the format/contents of the return dict.

cc @shuqiangzhang @kwen2501

shengbao-zheng commented:
@wconstab Thanks for pointing this out. Now I only dump similar PG related information captured by Execution Trace but missed in kineto profiler here, and they are currently enough for analysis and benchmarking purpose. Do you have any suggestion on the format/content?

aaronenyeshi reviewed (commented):
Overall LGTM, and thank you for adding updates to unit tests. I will let Will make the approval, after you've both resolved Will's comments

wconstab reviewed (commented):
I think you can land this as a non public API (put an underscore before the name). We can change the format later if we want. I suspect it's fine and we might only want to add new items to the dictionary but don't want to bind ourselves to it yet

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1066,6 +1066,30 @@ def get_backend(group: Optional[ProcessGroup] = None) -> Backend:
+    Return the pg configuration of all the process groups.
+
+    """
+    config_info: List[Dict[str, Union[int, str, List[int]]]] = []
+    for pg in _world.pg_map.keys():
```
The List return and the for loop here make me wonder if this function is actually doing "get-**all**-pg-configs", rather than what its name suggests.

We may want to consider if a "prime" function that's per group:
```
def _get_pg_config(group: Optional(ProcessGroup)=None)
```
would also get the job done.

kwen2501 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1203,11 +1203,21 @@ def test_3_level_hierarchical_model_averager(self):
+            real_group_ranks_res1 = []
+            real_group_ranks_res2 = []
+            pg_configs = dist.get_pg_config()
+            for pg_config in pg_configs:
+                if pg_configs['pg_name'] == subgroup1_name:
+                    real_group_ranks_res1 = pg_configs['ranks']
+                elif pg_configs['pg_name'] == subgroup2_name:
+                    real_group_ranks_res2 = pg_configs['ranks']
```
This for loop and if check seems to suggest that List-form return may not be the most ideal form.

kwen2501 reviewed (commented):
+1 on making it non-public at the moment.
Also added other comments.
Thanks!

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1066,6 +1066,41 @@ def get_backend(group: Optional[ProcessGroup] = None) -> Backend:
 
+def _get_pg_config(
+    group: Optional[ProcessGroup] = None,
+) -> Dict[str, Union[int, str, List[int]]]:
```
nit: might as well just use `Dict[str, Any]` to make it more readable.

shengbao-zheng committed:
[profiler] log process group config information in distributedInfo field (#119443)

Summary:

Process Group config is essential to analyze collective pattern. We have added this in Execution Trace. Now expose this information in Kineto as well

Reviewed By: aaronenyeshi, kwen2501

Differential Revision: D53557965
